docs/DevGuide/Parallelization.md

```diff
@@ -0,0 +1,542 @@
+the call to the Charm++ RTS for calls into array components. Charm++ refers to
+these types of remote calls as "inline entry methods". With the Charm++ method
+of eliding the RTS, the code becomes susceptible to stack overflows because
+of infinite recursion. The `receive_data` function is limited to at most 64 RTS
+elided calls, though in practice reaching this limit is rare. When the limit is
+reached the remote method invocation is done through the RTS instead of being
+elided.
+
+## 3. Reduction Actions {#dev_guide_parallelization_reduction_actions}
+
+Finally, there are reduction actions which are used when reducing data over an
+array. For example, you may want to know the sum of a `int` from every
+element in the array. You can do this as follows:
+
+\snippet Test_AlgorithmReduction.cpp contribute_to_reduction_example
+
+This reduces over the parallel component
+`ArrayParallelComponent<Metavariables>`, reduces to the parallel component
+`SingletonParallelComponent<Metavariables>`, and calls the action
+`ProcessReducedSumOfInts` after the reduction has been performed. The reduction
+action is:
+
+\snippet Test_AlgorithmReduction.cpp reduce_sum_int_action
+
+As you can see, the last argument to the `apply` function is of type `int`, and
+is the reduced value.
+
+You can also broadcast the result back to an array, even yourself. For example,
+
+\snippet Test_AlgorithmReduction.cpp contribute_to_broadcast_reduction
+
+It is often necessary to reduce custom data types, such as `std::vector` or
+`std::unordered_map`. Charm++ supports such custom reductions, and so does our
+layer on top of Charm++.
+Custom reductions require one additional step to calling
+`contribute_to_reduction`, which is writing a reduction function to reduce the
+custom data. We provide a generic type that can be used in custom reductions,
+`Parallel::ReductionData`, which takes a series of `Parallel::ReductionDatum` as
+template parameters and `ReductionDatum::value_type`s as the arguments to the
+constructor. Each `ReductionDatum` takes up to four template parameters (two
+are required). The first is the type of data to reduce, and the second is a
+binary invokable that is called at each step of the reduction to combine two
+messages. The last two template parameters are used after the reduction has
+completed. The third parameter is an n-ary invokable that is called once the
+reduction is complete, whose first argument is the result of the reduction. The
+additional arguments can be any `ReductionDatum::value_type` in the
+`ReductionData` that are before the current one. The fourth template parameter
+of `ReductionDatum` is used to specify which data should be passed. It is a
+`std::index_sequence` indexing into the `ReductionData`.
+
+The action that is invoked with the result of the reduction is:
+
+\snippet Test_AlgorithmReduction.cpp custom_reduction_action
+
+Note that it takes a `Parallel::ReductionData` object as its last argument.
```

Comment:
The snippet does not take a Parallel::ReductionData.  This isn't new in this PR, so it might be best to just file an issue to fix this section.